atom/browser/api/atom_api_content_tracing.cc

```diff
@@ -104,10 +116,15 @@ v8::Local<v8::Promise> StartTracing(
   atom::util::Promise promise(isolate);
   v8::Local<v8::Promise> handle = promise.GetHandle();
 
-  // Note: This method always succeeds.
-  TracingController::GetInstance()->StartTracing(
-      trace_config, base::BindOnce(atom::util::Promise::ResolveEmptyPromise,
-                                   std::move(promise)));
+  if (!TracingController::GetInstance()->StartTracing(
+          trace_config, base::BindOnce(atom::util::Promise::ResolveEmptyPromise,
+                                       std::move(promise)))) {
+    // If StartTracing returns false, that means it didn't invoke its callback.
+    // Return an already-resolved promise and abandon the previous promise (it
+    // was std::move()d into the StartTracing callback and has been deleted by
```

Comment:
Given the above, might it perhaps just be better to use a RepeatingCallback here?

Reply:
Not really; if I were going to use a repeating callback that would just force me to use `CopyablePromise`. But I could just as well use `CopyablePromise` with a `OnceCallback` to avoid the `std::move`. However, I think this is a better design, because it avoids unnecessary copying in the usual case—mostly, `startTracing` will only be called once, and this code won't be executed.